File: Labs/experiment2/D_swap_with_head.cpp
#include <iostream>
using namespace std;

int lst[1000005];
bool vis[1000005];

int main() {
    int n;
    cin >> n;
    for (int i = 1; i <= n; i++) cin >> lst[i];
    int ans = 0;
    for (int i = 1; i <= n; i++) {
        if (lst[i] == i || vis[i]) continue;
        int res = 0, tmp = i;
        if (i == 1) res--;
        else res++;
        while (!vis[tmp]) {
            res++;
            vis[tmp] = 1;
            tmp = lst[tmp];
        }
        ans += res;
    }
    cout << ans << endl;
    return 0;
}

File: Labs/experiment2/D_swap_with_head.py
n = int(input())
lst = [0] + list(map(int, input().split()))
vis = [False] * (n + 1)
ans = 0
for i in range(1, n + 1):
    if lst[i] == i or vis[i]:
        continue
    res = -1 if i == 1 else 1
    tmp = i
    while not vis[tmp]:
        res += 1
        vis[tmp] = True
        tmp = lst[tmp]
    ans += res
print(ans)
